lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -217,21 +217,51 @@ func (c *ContainerProxy) AddVolumesToContainer(handle string, config types.Conta
 	//Volume Attachment Section
 	log.Debugf("ContainerProxy.AddVolumesToContainer - VolumeSection")
 	log.Debugf("Raw Volume arguments : binds:  %#v : volumes : %#v", config.HostConfig.Binds, config.Config.Volumes)
-	var joinList []volumeFields
-	var err error
 
-	joinList, err = processAnonymousVolumes(config.Config.Volumes)
-	if err != nil {
-		return handle, derr.NewErrorWithStatusCode(fmt.Errorf("%s", err), http.StatusBadRequest)
+	// Collect all volume mappings. In a docker create/run, they
+	// can be anonymous (-v /dir) or specific (-v vol-name:/dir).
+	volumes := config.HostConfig.Binds
+	for anonVol := range config.Config.Volumes {
+		volumes = append(volumes, anonVol)
 	}
 
-	volumeSubset, err := processSpecifiedVolumes(config.HostConfig.Binds)
+	volList, err := processVolumeFields(volumes)
 	if err != nil {
-		return handle, derr.NewErrorWithStatusCode(fmt.Errorf("%s", err), http.StatusBadRequest)
+		return handle, derr.NewErrorWithStatusCode(fmt.Errorf("%s", err.Error()), http.StatusBadRequest)
 	}
-	joinList = append(joinList, volumeSubset...)
 
-	for _, fields := range joinList {
+	// Create and join volumes.
+	for _, fields := range volList {
+
+		driverArgs := make(map[string]string)
+		driverArgs["flags"] = fields.Flags
+
+		// NOTE: calling volumeCreate regardless of whether the volume is already
+		// present can be avoided by adding an extra optional param to VolumeJoin,
+		// which would then call volumeCreate if the volume does not exist.
+		vol := &Volume{}
+		_, req, err := vol.volumeCreate(fields.ID, "vsphere", driverArgs, nil)
```

Comment:
He either has to have request or he has to access the driverargs again when returning that error. However, the exact arguments passed through to the portlayer are in the request and could be scrubbed during the personality process before the call to the portlayer. The `req`'s values are the most accurate for what the failuer could have been.